src/core/jsproxy.c

```diff
@@ -644,26 +646,111 @@ JsProxy_Await(JsProxy* self, PyObject* _args)
   set_exception = _PyObject_GetAttrId(fut, &PyId_set_exception);
   FAIL_IF_NULL(set_exception);
 
-  JsRef promise_id = hiwire_resolve_promise(self->js);
-  JsRef idargs = hiwire_array();
-  JsRef idarg;
-  // TODO: does this leak set_result and set_exception? See #1006.
-  idarg = python2js(set_result);
-  hiwire_push_array(idargs, idarg);
-  hiwire_decref(idarg);
-  idarg = python2js(set_exception);
-  hiwire_push_array(idargs, idarg);
-  hiwire_decref(idarg);
-  hiwire_decref(hiwire_call_member(promise_id, "then", idargs));
-  hiwire_decref(promise_id);
-  hiwire_decref(idargs);
+  promise_id = hiwire_resolve_promise(self->js);
+  FAIL_IF_NULL(promise_id);
+  promise_handles = create_promise_handles(set_result, set_exception);
+  FAIL_IF_NULL(promise_handles);
+  promise_result = hiwire_call_member(promise_id, "then", promise_handles);
+  FAIL_IF_NULL(promise_result);
   result = _PyObject_CallMethodId(fut, &PyId___await__, NULL);
 
 finally:
   Py_CLEAR(loop);
+  Py_CLEAR(fut);
   Py_CLEAR(set_result);
   Py_CLEAR(set_exception);
-  Py_DECREF(fut);
+  hiwire_CLEAR(promise_id);
+  hiwire_CLEAR(promise_handles);
+  hiwire_CLEAR(promise_result);
+  return result;
+}
+
+/**
+ * Overload for `then` for JsProxies with a `then` method.
```

Comment:
What was the issue with the previous implementation? Memory leaks?

Reply:
Yeah, before we just allowed it to go through `JsProxy_Vectorcall`. This leaked the `onFulfilled` and `onResolved` handles, and you can't fix it with `create_once_proxy` because exactly one of `onFulfilled` and `onResolved` gets called. Actually the implementation of `catch` is still wrong though.